File: java/src/phase_field_model/Cell.java
package phase_field_model;

public class Cell implements Field2D {

  private int lx, ly;
  private double volume;

  // Store cm coordinates in cell's own reference frame
  private double xcm, ycm;
  private double deltaXCM, deltaYCM;

  private int cellType;
  private int x, y;
  private double cellField[][][];
  private int setField, getField; // For switching between old and new field
  private static double INCELL = 0.4;
  private static double CMSHIFT = 2.0;
  private VolumeField volumeField;

  public Cell(int x, int y, int lx, int ly, int type) {
    this.x = x;
    this.y = y;
    this.lx = lx;
    this.ly = ly;
    this.xcm = lx / 2.0;
    this.ycm = ly / 2.0;
    this.deltaXCM = 0.0;
    this.deltaYCM = 0.0;
    this.cellType = type;
    this.cellField = new double[2][lx][ly]; // Default field value is zero
    this.setField = 1;
    this.getField = 1;
    this.volumeField = new VolumeField(this);
  }

  public void initSquareCell(int dx, int dy){
    int xStart = (int) (lx/2.0 - dx/2.0);
    int xEnd = xStart + dx;
    int yStart = (int) (ly/2.0 - dy/2.0);
    int yEnd = yStart + dy;
    for (int k = 0; k < 2; k++){
      for (int i = xStart; i < xEnd; i++){
        for (int j = yStart; j < yEnd; j++){
          cellField[k][i][j] = 1.0;
        }
      }
    }
    double [] cm = calculateCM();
    xcm = cm[0];
    ycm = cm[1];
  }
  
  public void initCell(double [][] matrix){
    if (matrix.length == lx && matrix[0].length == ly){
      for (int i = 0; i < lx; i++){
        for (int j = 0; j < ly; j++){
          cellField[0][i][j] = matrix[i][j];
          cellField[1][i][j] = matrix[i][j];
        }
      }
    }
    double [] cm = calculateCM();
    xcm = cm[0];
    ycm = cm[1];
  }

  public void updateTotalVolume() {
    volume = calculateTotalVolume();
  }
  
  protected double calculateTotalVolume() {
    double sum = 0.0;
    for (int i = 0; i < lx; i++) {
      for (int j = 0; j < ly; j++) {
        sum += volumeField.get(i, j);
      }
    }
    return sum;
  }
  
  protected double [] calculateCM() {
    double xavg = 0;
    double yavg = 0;
    int count = 0;
    for (int i = 0; i < lx; i++) {
      for (int j = 0; j < ly; j++) {
        if (cellField[getField][i][j] > INCELL) {
          xavg += i;
          yavg += j;
          count++;
        }
      }
    }
    if (count > 0){
      xavg /= (double) count;
      yavg /= (double) count;
    } else {
      xavg = 0.0;
      yavg = 0.0;
    }
    return new double [] {xavg, yavg};
  }

  public void updateCM() {
	  
    double oldXCM = xcm;
    double oldYCM = ycm;
    
    double [] cm = calculateCM();
    xcm = cm[0];
    ycm = cm[1];

    deltaXCM += (xcm - oldXCM);
    deltaYCM += (ycm - oldYCM);
    
    if (Math.abs(deltaXCM) > CMSHIFT || Math.abs(deltaYCM) > CMSHIFT) {
      int xShift = (int) Math.floor(deltaXCM);
      int yShift = (int) Math.floor(deltaYCM);
      shiftCoordinates(xShift, yShift);
      deltaXCM = 0;
      deltaYCM = 0;
      cm = calculateCM();
      xcm = cm[0];
      ycm = cm[1];
    }
  }

  public void shiftCoordinates(int xShift, int yShift) {
    startUpdateCellField();
    int xStart, xEnd, yStart, yEnd;
    int zeroXStart, zeroXEnd, zeroYStart, zeroYEnd;
    if (xShift >= 0) {
      xStart = xShift;
      xEnd = lx;
      zeroXStart = lx - xShift;
      zeroXEnd = lx;
    } else {
      xStart = 0;
      xEnd = lx + xShift;
      zeroXStart = 0;
      zeroXEnd = -xShift;
    }
    if (yShift >= 0) {
      yStart = yShift;
      yEnd = ly;
      zeroYStart = ly - yShift;
      zeroYEnd = ly;
    } else {
      yStart = 0;
      yEnd = ly + yShift;
      zeroYStart = 0;
      zeroYEnd = -yShift;
    }
    
    // Set empty cells to zero
    for (int i = zeroXStart; i < zeroXEnd; i++) {
      for (int j = 0; j < ly; j++) {
        set(i, j, 0.0);
      }
    }
    for (int i = zeroXEnd; i < lx; i++) {
      for (int j = zeroYStart; j < zeroYEnd; j++) {
        set(i, j, 0.0);
      }
    }
    
    // Shift cells
    for (int i = xStart; i < xEnd; i++) {
      for (int j = yStart; j < yEnd; j++) {
        set(i - xShift, j - yShift, get(i, j));
      }
    }
    
    // Update end coordinate
    x += xShift;  
    y += yShift;
    endUpdateCellField();
  }

  // Accessor methods
  public double getXCM() {
    return xcm;
  }

  public double getYCM() {
    return ycm;
  }

  public int getX() {
    return x;
  }

  public int getY() {
    return y;
  }
  
  public double getVolume(int i, int j) {
    return volumeField.get(i, j);
  }

  public double getTotalVolume() {
    return volume;
  }

  public Field2D getVolumeField() {
    return volumeField;
  }

  public int getCellType() {
    return cellType;
  }

  @Override
  public int getLx() {
    return lx;
  }

  @Override
  public int getLy() {
    return ly;
  }

  @Override
  public double get(int i, int j) {
    return cellField[getField][i][j];
  }

  @Override
  public void set(int i, int j, double value) {
    cellField[setField][i][j] = value;
  }
  
  protected void startUpdateCellField() {
    setField = getField == 1 ? 0 : 1;
  }

  protected void endUpdateCellField() {
    getField = setField;
  }
}

File: c/nonomura_model/cell.c
// cell.c

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include "cell.h"
#include "array.h"
#include "mtwister.h"

void initCell(Cell* cell, int _x, int _y, int _lx, int _ly, int _t, int seed) {
  cell->x = _x;
  cell->y = _y;
  cell->lx = _lx;
  cell->ly = _ly;
  cell->type = _t;

  // Allocate memory and initialise the field to zero
  for (int i = 0; i < 2; i++) {
    cell->field[i] = create2DDoubleArray(cell->lx, cell->ly);
  }
  cell->volumeField = create2DDoubleArray(cell->lx, cell->ly);
  cell->setIndex = 1;
  cell->getIndex = 1;
  cell->vx = 0.0;
  cell->vy = 0.0;
  cell->theta = 0.0;
  cell->rotateDiff = 0.01;
  cell->xcm = cell->lx/2.0;
  cell->ycm = cell->ly/2.0;
  cell->deltaXCM = 0.0;
  cell->deltaYCM = 0.0;
  cell->volume = 0.0;
  srand(seed);
  cell->random = seedRand(rand());
}

void deleteCell(Cell* cell) {
  for (int i = 0; i < 2; i++) {
    free(cell->field[i]);
  }
  free(cell->volumeField);
  free(cell);
}

void initFieldSquareOnes(Cell* cell, int x0, int y0, int dx, int dy) {
  for (int i = 0; i < dx && x0+i < cell->lx-1; i++) {
    for (int j = 0; j < dy && y0+j < cell->ly-1; j++) {
      cell->field[0][i+x0][j+y0] = 1.0;
      cell->field[1][i+x0][j+y0] = 1.0;
    }
  }
  calculateCM(cell, &cell->xcm, &cell->ycm);
  updateVolume(cell);
}

void calculateCM(Cell* cell, double* xcm, double* ycm) {
  double xavg = 0.0;
  double yavg = 0.0;
  int count = 0;
  for (int i = 0; i < cell->lx; i++) {
    for (int j = 0; j < cell->ly; j++) {
      if (cell->field[cell->getIndex][i][j] > INCELL) {
	xavg += i;
	yavg += j;
	count++;
      }
    }
  }
  if (count > 0) {
    xavg /= (double) count;
    yavg /= (double) count;
  } else {
    xavg = 0.0;
    yavg = 0.0;
  }
  *xcm = xavg;
  *ycm = yavg;
}

void updateCM(Cell* cell) {
  double oldXCM = cell->xcm;
  double oldYCM = cell->ycm;
  calculateCM(cell, &cell->xcm, &cell->ycm);
  cell->deltaXCM += (cell->xcm - oldXCM);
  cell->deltaYCM += (cell->ycm - oldYCM);

  if (fabs(cell->deltaXCM) > CMSHIFT || fabs(cell->deltaYCM) > CMSHIFT) {
    int xShift = (int)(floor(cell->deltaXCM));
    int yShift = (int)(floor(cell->deltaYCM));
    shiftCoordinates(cell, xShift, yShift);
    cell->deltaXCM -= xShift;
    cell->deltaYCM -= yShift;
    //    calculateCM(cell, &cell->xcm, &cell->ycm);
    cell->xcm -= xShift;
    cell->ycm -= yShift;
  }
}

void updateVolume(Cell* cell) {
  double totalVolume = 0.0;
  int get = cell->getIndex;
  for (int i = 0; i < cell->lx; i++) {
    for (int j = 0; j < cell->ly; j++) {
      double u = cell->field[get][i][j];
      cell->volumeField[i][j] = u*u*(3.0-2.0*u);
      if (cell->volumeField[i][j] > INCELL) {
	totalVolume += 1.0;
      }
    }
  }
  cell->volume = totalVolume;
}

void updateVelocity(Cell* cell) {
  cell->theta += sqrt(4.0 * cell->rotateDiff) *
    (genRand(&cell->random)*2.0-1.0);
  cell->vx = cos(cell->theta);
  cell->vy = sin(cell->theta);
}

void shiftCoordinates(Cell* cell, int xShift, int yShift) {
  startUpdateCellField(cell);
  int set = cell->setIndex;
  int get = cell->getIndex;
  int xStart, xEnd, yStart, yEnd;
  int zeroXStart, zeroXEnd, zeroYStart, zeroYEnd;
  if (xShift >= 0) {
    xStart = xShift;
    xEnd = cell->lx;
    zeroXStart = cell->lx - xShift;
    zeroXEnd = cell->lx;
  } else {
    xStart = 0;
    xEnd = cell->lx + xShift;
    zeroXStart = 0;
    zeroXEnd = -xShift;
  }
  if (yShift >= 0) {
    yStart = yShift;
    yEnd = cell->ly;
    zeroYStart = cell->ly - yShift;
    zeroYEnd = cell->ly;
  } else {
    yStart = 0;
    yEnd = cell->ly + yShift;
    zeroYStart = 0;
    zeroYEnd = -yShift;
  }

  // Set empty cells to zero
  for (int i = zeroXStart; i < zeroXEnd; i++) {
    for (int j = 0; j < cell->ly; j++) {
      cell->field[set][i][j] = 0.0;
    }
  }
  for (int i = zeroXEnd; i < cell->lx; i++) {
    for (int j = zeroYStart; j < zeroYEnd; j++) {
      cell->field[set][i][j] = 0.0;
    }
  }

  // Shift cells
  for (int i = xStart; i < xEnd; i++) {
    for (int j = yStart; j < yEnd; j++) {
      cell->field[set][i-xShift][j-yShift] = cell->field[get][i][j];
    }
  }

  // Update end coordinate
  cell->x += xShift;
  cell->y += yShift;
  endUpdateCellField(cell);
}

void startUpdateCellField(Cell* cell) {
  cell->setIndex = (cell->getIndex == 1 ? 0 : 1);
}

void endUpdateCellField(Cell* cell) {
  cell->getIndex = cell->setIndex;
}
